contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,215 @@
+            return None;
+        }
+    };
+
+    let connection_config = match database_configs
+        .iter()
+        .filter(|e| {
+            match e.as_table().map(|table| {
+                table.get("name").map(|v| v.as_str()) == Some(Some(name))
+            }) {
+                Some(true) => true,
+                _ => false,
+            }
+        })
+        .nth(0) {
+        Some(config) => config,
+        None => {
+            error_!("Could not find any database configurations named: {}", name);
+            warn_!("Available configurations: {}", available_database_configs(from).join(", "));
+            return None;
+        }
+    };
+
+    let url = match connection_config.get("url") {
+        Some(url) => {
+            match url.as_str() {
+                Some(url) => url,
+                None => {
+                    error_!("Could not convert the URL into a valid string");
+                    return None;
+                }
+            }
+        }
+        None => {
+            error_!("Could not fetch the URL as part of the connection config");
+            return None;
+        }
+    };
+
+    let pool_size = match connection_config.get("pool_size") {
+        Some(pool_size) => {
+            pool_size.as_integer().unwrap_or(from.workers as i64) as u32
+        }
+        None => from.workers as u32,
+    };
+
+    let extras = connection_config.as_table();
+
+    Some(DatabaseConfig {
+        url,
+        pool_size,
+        extras,
+    })
+}
+
+/// Returns an iterator containing the names of all the possible database connections.
+/// The primary use within `rocket_contrib` for this method is in relation to calling out
+/// valid database connections that are specified in the `Rocket.toml` file.
+pub fn available_database_configs<'a>(config: &'a config::Config) -> Vec<&'a str> {
+    let db_config = match config.get_slice("databases") {
+        Ok(config) => config,
+        Err(_) => return vec![],
+    };
+
+    db_config
+        .iter()
+        .map(|config| {
+            config.as_table().and_then(|table| table["name"].as_str())
+        })
+        .filter_map(|e| e)
+        .collect::<Vec<&'a str>>()
+}
+
+/// This trait can be implemented on Connection types that can be stored in a connection pool.
+/// Implementing this trait on a type allows Rocket to generate the necessary code to manage the
+/// pool internally and inject it into request handlers.
+pub trait Poolable: Send + Sized + 'static {
+    type Manager: ManageConnection<Connection=Self>;
+    type Error;
+
+    fn pool(config: DatabaseConfig) -> Result<r2d2::Pool<Self::Manager>, Self::Error>;
+}
+
+#[cfg(feature = "diesel_sqlite")]
+impl Poolable for diesel::SqliteConnection {
+    type Manager = diesel::r2d2::ConnectionManager<diesel::SqliteConnection>;
+    type Error = DbError<!>;
+
+    fn pool(config: DatabaseConfig) -> Result<r2d2::Pool<<Self as Poolable>::Manager>, <Self as Poolable>::Error> {
```

Comment:
There should be a need to do `Self as Poolable` here. Use the signature from the trait directly:

```rust
fn pool(config: DatabaseConfig) -> Result<r2d2::Pool<Self::Manager>, Self::Error>;
```

Reply:
Yeah, I think Intellj's Rust plugin is what used that type as such. Will fix it up.